File: bridge/ts/packages/bridge-types/src/envelope.ts
// Bridge wire envelope types.
// Mirrors envelope.rs Envelope struct and Kind enum exactly.
// Field names follow the JSON wire format (camelCase where Rust uses rename).

export const KINDS = ['request', 'response', 'event'] as const;
export type Kind = (typeof KINDS)[number];

/**
 * Wire error object carried in a response envelope's error field.
 * Mirrors error.rs BridgeError (additionalProperties: false).
 */
export interface BridgeError {
  /** Screaming-snake-case error code, e.g. "METHOD_NOT_SUPPORTED". */
  code: string;
  /** Human-readable error message. */
  message: string;
  /** Optional structured, error-code-specific detail. */
  data?: Record<string, unknown>;
}

/**
 * Canonical Bridge wire envelope.
 * Mirrors envelope.rs Envelope with serde field names (camelCase: sessionId).
 * Optional fields are present/absent per-kind as enforced by Envelope::validate.
 */
export interface Envelope {
  /** Protocol marker — always "norves.editor.bridge". */
  bridge: 'norves.editor.bridge';
  /** Protocol version string, MAJOR.MINOR. */
  version: string;
  /** Envelope discriminator. */
  kind: Kind;
  /** Request/response correlation id. Present on request and response. */
  id?: string;
  /** Method name on a request. */
  method?: string;
  /** Event name on an event envelope. */
  event?: string;
  /** Method or event payload. */
  params?: Record<string, unknown>;
  /** Success payload on a response. Mutually exclusive with error. */
  result?: unknown;
  /** Error payload on a response. Mutually exclusive with result. */
  error?: BridgeError;
  /** Optional session id assigned during the handshake. Wire name: sessionId. */
  sessionId?: string;
  /** Optional monotonically increasing per-connection sequence number. */
  seq?: number;
}

File: bridge/cpp/engine-sdk/src/envelope.cpp
#include "norves/bridge/envelope.hpp"

#include "norves/bridge/codec_error.hpp"
#include "norves/bridge/result.hpp"

#include <variant>

// validate() は Rust リファレンスの Envelope::validate
// （bridge/crates/norves-bridge-core/src/envelope.rs, fn validate）の 1:1 移植である。
// 2 つの実装が同じ最初の違反を報告するよう、チェックの順序と違反メッセージは同一に
// 保たれている。
namespace norves::bridge
{

    namespace
    {

        Result<std::monostate, CodecError> Violation(const char* message)
        {
            return Result<std::monostate, CodecError>::err(
                CodecError::structural_violation(message));
        }

        Result<std::monostate, CodecError> Ok()
        {
            return Result<std::monostate, CodecError>::ok(std::monostate{});
        }

    }  // namespace

    Result<std::monostate, CodecError> Envelope::validate() const
    {
        switch (kind)
        {
            case Kind::Request:
                if (!id.has_value())
                {
                    return Violation("request envelope requires `id`");
                }
                if (!method.has_value())
                {
                    return Violation("request envelope requires `method`");
                }
                if (result.has_value())
                {
                    return Violation("request envelope must not carry `result`");
                }
                if (error.has_value())
                {
                    return Violation("request envelope must not carry `error`");
                }
                if (event.has_value())
                {
                    return Violation("request envelope must not carry `event`");
                }
                break;
            case Kind::Response:
            {
                if (!id.has_value())
                {
                    return Violation("response envelope requires `id`");
                }
                const bool bHasResult = result.has_value();
                const bool bHasError = error.has_value();
                if (bHasResult && bHasError)
                {
                    return Violation("response envelope must not carry both `result` and `error`");
                }
                if (!bHasResult && !bHasError)
                {
                    return Violation(
                        "response envelope requires exactly one of `result` or `error`");
                }
                if (method.has_value())
                {
                    return Violation("response envelope must not carry `method`");
                }
                if (event.has_value())
                {
                    return Violation("response envelope must not carry `event`");
                }
                if (params.has_value())
                {
                    return Violation("response envelope must not carry `params`");
                }
                break;
            }
            case Kind::Event:
                if (!event.has_value())
                {
                    return Violation("event envelope requires `event`");
                }
                if (id.has_value())
                {
                    return Violation("event envelope must not carry `id`");
                }
                if (method.has_value())
                {
                    return Violation("event envelope must not carry `method`");
                }
                if (result.has_value())
                {
                    return Violation("event envelope must not carry `result`");
                }
                if (error.has_value())
                {
                    return Violation("event envelope must not carry `error`");
                }
                break;
        }
        return Ok();
    }

}  // namespace norves::bridge
